### PR TITLE
New linter API

### PR DESCRIPTION
**Pull request by remiq**
Solves issues #8 #11 #13

### CODE REVIEW

File: lib/init.coffee [hunk 1]
```diff
@@ -1,14 +1,67 @@
...
+      (executablePath) =>
+        @executablePath = executablePath
+  deactivate: ->
+    @subscriptions.dispose()
+  provideLinter: ->
+    provider =
+      grammarScopes: ['source.elixir']
+      scope: 'file'
+      lintOnFly: false
+      name: 'Atom-Elixir'
```
Comment (steelbrain) [inline lib/init.coffee]: How about just `Elixir`?

File: lib/init.coffee [hunk 2]
```diff
@@ -1,14 +1,67 @@
...
+      lint: (textEditor) =>
+        return new Promise (resolve, reject) =>
+          error_stack = []
+          compile_result = ''
+          helpers = require('atom-linter')
```
Comment (steelbrain) [inline lib/init.coffee]: Can you move these requires one level up?
I don't want them to be imported on every lint call. They stay the same.

File: lib/init.coffee [hunk 3]
```diff
@@ -1,14 +1,67 @@
...
+            re_result = re.exec(row)
+            ret =
+              type: re_result[1]
+              text: re_result[4]
+              fp: filePath
+              filePath: project_path[0] + '/' + re_result[2]
+              range: helpers.rangeFromLineNumber(textEditor, re_result[3] - 1)
+            #console.log ret
+            error_stack.push ret
+          process = new BufferedProcess
```
Comment (steelbrain) [inline lib/init.coffee]: We have `atom-linter::exec` in that module

**Comment by remiq** [comment]
Changed name to Elixir
Moved requires one level up
Used atom-linter::exec instead of writing my own.

File: lib/init.coffee [hunk 4]
```diff
@@ -1,14 +1,52 @@
...
+      name: 'Elixir'
+      lint: (textEditor) =>
+        filePath = textEditor.getPath()
+        elixirc_args = [
+          "--ignore-module-conflict", "--app", "mix", "--app", "ex_unit", "-o", os.tmpDir(),
+          filePath
+        ]
+        project_path = atom.project.getPaths()
+        opts =
+          cwd: project_path[0]
```
Comment (steelbrain) [inline lib/init.coffee]: How about you require `path` where you required `os` and then do `path.dirname(filePath)`

File: lib/init.coffee [hunk 5]
```diff
@@ -1,14 +1,52 @@
...
+        helpers.exec(@executablePath, elixirc_args, opts)
+          .then (compile_result) ->
+            error_stack = []
+            parse_row = (row) ->
```
Comment (steelbrain) [inline lib/init.coffee]: If the `parse_row` function doesn't consume anything from this scope, can we make it out on the top level.

**Comment by steelbrain** [comment]
You're doing a great job by the way :+1: 
Also, you can remove the outdated file `linter-x.coffee` from this repo now.

**Comment by remiq** [comment]
Using `path.dirname(filePath)` instead `project_path[0]`
`parse_row` function no longer alters `error_stack`, instead returns one warning per line.

File: lib/init.coffee [hunk 6]
```diff
@@ -1,14 +1,50 @@
...
+      lintOnFly: false
+      name: 'Elixir'
+      lint: (textEditor) =>
+        filePath = textEditor.getPath()
+        filePathDir = path.dirname(filePath)
+        elixirc_args = [
+          "--ignore-module-conflict", "--app", "mix", "--app", "ex_unit", "-o", os.tmpDir(),
+          filePath
+        ]
+        project_path = atom.project.getPaths()
```
Comment (steelbrain) [inline lib/init.coffee]: This variable is un-used

**Comment by steelbrain** [comment]
Looks pretty good, now all we need is a person confirming that it works.

File: lib/init.coffee [hunk 7]
```diff
@@ -1,14 +1,49 @@
...
-      default: ''
-    includeDirs:
-      type: 'string'
-      default: ''
-    pa:
+    executablePath:
       type: 'string'
-      default: ''
+      title: 'Elixirc path'
+      default: '/usr/local/bin/elixirc'
```
Comment (Arcanemagus) [inline lib/init.coffee]: You might want to default this to just `elixirc` and let the user's environment figure out the real location on their system.

**Comment by Arcanemagus** [comment]
You should probably also ensure the `linter` package is installed for your users by using the `atom-package-deps` package. To configure that you specify any desired Atom packages in your `package.json` (like [this](https://github.com/AtomLinter/linter-php/blob/master/package.json#L15-L17)) and then calling atom-package-deps.install (like [this](https://github.com/AtomLinter/linter-php/blob/master/lib/main.coffee#L11)).

**Comment by remiq** [comment]
Warnings are handled.
Using CoffeeScript multiline regexps.
Using `atom-package-deps` to install `linter` pakage.
Default `elixirc` path is `elixirc`.

**Comment by Arcanemagus** [comment]
@remiq This seems to be working fine for me on Ubuntu, I'm not going to bother compiling this on Windows but it should work equally well there. You should probably update the README.md file as well to reflect the new state of the plugin. Thanks for your hard work!

@steelbrain If this looks good to you I would say a team should be created and @remiq become the new maintainer of this package, assuming you have no issues with that @remiq :wink:.

**Comment by steelbrain** [comment]
Sure, waiting on the green signal from him

**Comment by remiq** [comment]
Ok, I guess I could do that.

**Comment by steelbrain** [comment]
You should've recieved emails from GitHub

**Comment by remiq** [comment]
Linter now detects if project directory has `mix.exs` file. 
If yes, it will use `mix compile`, which will solve #9, #10, #12.
If no, it will use `elixirc` for single file.

Invitation received and accepted.

**Comment by Arcanemagus** [comment]
@remiq welcome to the club :stuck_out_tongue:

A few pointers if you haven't published a package on Atom before:
You will need to publish new versions by checking out this base repository somewhere, then running `apm publish {major|minor|patch}` in that directory. It will create a new tag and modify the version in `package.json` for you. If you want GitHub to also show which is the latest release  you will need to go to the [releases](https://github.com/AtomLinter/linter-elixirc/releases) and edit the latest tag to mark it as latest on GitHub.

Since this is a complete re-write I would do a major release (v1.0.0).

**Pull request merged**